unicorn/include/cocaine/zookeeper/handler.hpp

```diff
@@ -0,0 +1,230 @@
+
+    handler_dispatcher_t() {}
+    handler_dispatcher_t(const handler_dispatcher_t& other) = delete;
+    handler_dispatcher_t& operator=(const handler_dispatcher_t& other) = delete;
+    std::mutex storage_lock;
+    storage_t callbacks;
+};
+
+/**
+* Scope of the handlers. After this goes out of scope all handlers ae destroyed.
+* However call of this handlers via handler manager is defined and do nothing.
+* Thread unsafe. For thread safety consider external locking or creating separate class.
+*/
+class handler_scope_t {
+public:
+    handler_scope_t() :
+        registered_callbacks()
+    {}
+
+    ~handler_scope_t();
+
+    template<class T, class ...Args>
+    T&
+    get_handler(Args&& ...args) {
+        auto handler = new T(handler_tag(), std::forward<Args>(args)...);
+        handler_dispatcher_t::instance().add(handler);
+        registered_callbacks.push_back(handler);
+        return *handler;
+    }
+
+private:
+    std::vector<managed_handler_base_t*> registered_callbacks;
+};
+
+class managed_watch_handler_base_t :
+    virtual public managed_handler_base_t
+{
+public:
+    virtual void
+        operator() (int type, int state, path_t path) = 0;
+
+    managed_watch_handler_base_t(const handler_tag& tag) :
+        managed_handler_base_t(tag)
+    {}
+};
+
+/**
+* At this time managed void handler is unneeded so it's being deleted after invocation.
+* Maybe better to stick to only managed handlers for consistency.
+*/
+class void_handler_base_t {
+public:
+    virtual void
+    operator() (int rc) = 0;
+
+    virtual
+    ~void_handler_base_t() {}
+};
+
+class managed_stat_handler_base_t :
+    virtual public managed_handler_base_t
+{
+public:
+    virtual void
+    operator() (int rc, const node_stat& stat) = 0;
+
+    managed_stat_handler_base_t(const handler_tag& tag) :
+        managed_handler_base_t(tag)
+    {}
+
+    virtual
+    ~managed_stat_handler_base_t() {}
+};
+
+class managed_data_handler_base_t :
+    virtual public managed_handler_base_t
+{
+public:
+    virtual void
+    operator() (int rc, value_t value, const node_stat& stat) = 0;
+
+    managed_data_handler_base_t(const handler_tag& tag) :
+        managed_handler_base_t(tag)
+    {}
+
+    virtual
+    ~managed_data_handler_base_t() {}
+};
+
+class managed_string_handler_base_t :
+    virtual public managed_handler_base_t
+{
+public:
+    virtual void
+    operator() (int rc, value_t value) = 0;
+
+    managed_string_handler_base_t(const handler_tag& tag) :
+        managed_handler_base_t(tag)
+    {}
+
+    virtual
+    ~managed_string_handler_base_t() {}
+};
+
+class managed_strings_stat_handler_base_t:
+    virtual public managed_handler_base_t
+{
+public:
+    virtual void
+        operator() (int rc, std::vector<std::string> childs, const node_stat& stat) = 0;
+
+    managed_strings_stat_handler_base_t(const handler_tag& tag) :
+        managed_handler_base_t(tag)
+    {}
+};
+
+
+typedef std::unique_ptr<void_handler_base_t> void_handler_ptr;
```

Comment:
IMO with this typedef you hide the ownership, a contract. Whatever, since nobody else uses this typedef.